zstd/zstd.go

```diff
@@ -27,14 +27,30 @@ import (
 
 const Name = "zstd"
 
+var encoderOptions = []zstd.EOption{
+	// The default zstd window size is 8MB, which is much larger than the
+	// typical RPC message and wastes a bunch of memory.
+	zstd.WithWindowSize(512 * 1024),
+	// By default zstd uses GOMAXPROCS (64 in prod) threads to encode a
```

Comment:
@jaronoff97, this statement is true in Lightstep's version of this package but in this file where EncodeAll is used I don't think it's relevant because EncodeAll always uses a single goroutine. Probably only the window size setting is necessary here. (You should double-check those things but I'm pretty sure.)

Reply:
It seems if its unset, i think it will use however many CPUs you have https://github.com/klauspost/compress/blob/master/zstd/encoder_options.go#L36

Reply:
> It seems if its unset, i think it will use however many CPUs you have https://github.com/klauspost/compress/blob/master/zstd/encoder_options.go#L36

"// This function [EncodeAll] can be called concurrently, but each call will only run on a single goroutine."

https://github.com/klauspost/compress/blob/9a951bee0e5e9df261e6ca6a68bfa1e819f333d0/zstd/encoder.go#L469

Again, this is a difference compared to Lightstep's package which doesn't use EncodeAll.

Reply:
I think the one added is the main one – the others are pretty lightstep specific :)

Reply:
> Just curious- do you have other changes worth upstreaming?

One other possibility: instead of this package blindly registering its compressor as

```
func init() {
    ...
    encoding.RegisterCompressor(c)
}
```

you might consider a small guard in order to be slightly friendlier to applications that are already using their own compressors under the same name:

```
func init() {
    ...
    if encoding.GetCompressor("zstd") == nil {
        encoding.RegisterCompressor(c)
    }
}
```

I.e. you don't install over something that's already there.

This is actually how we ended up looking at this package: it's being linked into the OpenTelemetry SDK, so when that SDK started getting linked into some of our binaries we noticed that our zstd memory went through the roof. The initialization order was simply that our existing registration was going first and yours happened to be going second and clobbering ours. The new defaults that jaronoff97 put in here are friendlier in general but the entire situation would have been avoided if your registration had a little guard like the one above.

(Although it's debatable whether such a guard is actually your responsibility or not. I sort of think the OpenTelemetry SDK should be a little more careful itself about this kind of accidental clobbering, but 🤷.)

Reply:
Interesting. The use of `init()` isn't great, it's something I copied from the grpc code :/

Reply:
> I'm not sure that adding a check like this would remove the race condition though, wouldn't it just invert the result? ie first compressor to register with a certain name wins, rather than the last?

No, it means that whether or not your package goes first, my package wins. (My package wouldn't have the guard, only yours.) Which is exactly as it should be: I want to be able to control which package my company uses and I don't want that choice to change just because yours gets accidentally linked in through some long dependency chain that is obscure to both of us.

Although, again, it's not clear to me that this is necessarily your problem. I was just trying to answer your question upthread about other things that might be helpful. I sort of think that the problem here is ultimately with the OpenTelemetry SDK: _they_ should be good citizens and not do anything that might clobber their users.

Reply:
> No, it means that whether or not your package goes first, my package wins. (My package wouldn't have the guard, only yours.) Which is exactly as it should be: I want to be able to control which package my company uses and I don't want that choice to change just because yours gets accidentally linked in through some long dependency chain that is obscure to both of us.

Ah, I see. I was imagining the scenario where both packages had the guard.

> Although, again, it's not clear to me that this is necessarily your problem. I was just trying to answer your question upthread about other things that might be helpful. I sort of think that the problem here is ultimately with the OpenTelemetry SDK: _they_ should be good citizens and not do anything that might clobber their users.

Is there anything the OpenTelemetry SDK could have done to avoid this though, besides not using this package? Your solution seems reasonable, but I wonder if is there anything more flexible/controllable by the main package?